testfun.hpp

```diff
@@ -0,0 +1,130 @@
+    {
+        std::cout<<"Evaluate Not Implemented";
+        std::terminate();
+        return 0;
+    };
+};
+
+// Quadratic function x^Tx, gradient implemented exactly. Global min at x=0
+class Quadratic: public BaseObjective{
+public:
+    Quadratic(const int d){}
+    double evaluate(const Eigen::VectorXd& x)
+    {
+        // In 2D, x^2+y^2 (or x^T * x)
+        return x.squaredNorm();
+    }
+    Eigen::VectorXd gradient(const Eigen::VectorXd& x)
+    {
+        return 2*x;
+    }
+};
+
+// 1 dimension double well, overloads with exact gradient. Global min x = +-1/sqrt(2)
+class DoubleWell1D: public BaseObjective{
+public:
+    DoubleWell1D(const int d){assert(d==1 && "Only well-defined for 1 dimension");}
+    double evaluate(const Eigen::VectorXd& x)
+    {
+        // x^4 - x^2, .sum() is a hack to return a double instead of vector
+        return (x.array().pow(4) - x.array().square()).sum();
+    }
+    Eigen::VectorXd gradient(const Eigen::VectorXd& x)
+    {
+        // 4x^3 - 2x
+        return 4*x.array().pow(3) - 2*x.array();
+    }
+};
+
+// Double well with two minima, global min at x = (3+sqrt(41))/8 ~ 1.1754
+class UnevenDoubleWell1D: public BaseObjective{
+public:
+    UnevenDoubleWell1D(const int d){assert(d==1 && "Only well-defined for 1 dimension");}
+    double evaluate(const Eigen::VectorXd& x)
+    {
+        // x^4 -x^3 - x^2  .sum() returns double instead of Vector
+        return (x.array().pow(4) - x.array().cube() - x.array().square()).sum();
+    }
+    Eigen::VectorXd gradient(const Eigen::VectorXd& x)
+    {
+        // 4x^3 - 3x^2 - 2x
+        return 4*x.array().pow(3) - 3*x.array().square() - 2*x.array();
+    }
+};
+
+// Eggholder function, unknown global minimum
+class Eggholder: public BaseObjective{
+public:
+    Eggholder(const int d){assert(d==2 && "Eggholder only defined for d==2");}
+    double evaluate(const Eigen::VectorXd& x)
+    {
+        return -(x(1)+47)*sin(sqrt(abs((x(0)/2) + x(1)+47))) - x(0)*sin(sqrt(abs(x(0) - x(1) - 47)));
+    }
+};
+
+// Shekel function
+// Needs restricting to  [-10,10] where minimum is at 4,4,4,4
+// Something is incorrect in the implementation
+class Shekel: public BaseObjective{
+public:
+    Shekel(const int d){assert(d==4 && "Shekel only defined for d==4");}
+    double evaluate(const Eigen::VectorXd& x)
+    {
+        Eigen::MatrixXd C(4,10);
+        Eigen::MatrixXd  beta(10,1);
+        C <<  4, 1, 8, 6, 3, 2, 5, 8, 6, 7,
+              4, 1, 8, 6, 7, 9, 3, 1, 2, 3.6,
+              4, 1, 8, 6, 3, 2, 5, 8, 6, 7,
+              4 ,1, 8, 6, 7, 9, 3, 1, 2, 3.6;
+        beta << 1, 2, 2, 4, 4, 6, 3, 7, 5, 5;
+        beta /= 10;
+        double val = 0;
+        for(int i=0; i<x.size();++i)
+        {
+            for(int j=0; j<4;++j)
+            {
+                val += (pow((x(j) - C(j,i)),2));
+            }
+            val += beta(i);
+            val = 1/val;
+        }
+        val = -val;
+        return val;
```

Comment:
I think I may have spotted your bug here. You are doing val =1/val for every iteration over j, then adding to val on the next iteration and inverting the whole thing again, i.e. a^{-1} + b^{-1} \neq (a^{-1} + b )^{-1}. Maybe something like this would work?
double val = 0;
double buffer = 0;
        for(int i=0; i<x.size();++i)
        {
            buffer = 0;
            for(int j=0; j<4;++j)
            {
                buffer += (pow((x(j) - C(j,i)),2));
            }
            buffer += beta(i);
            buffer = 1/buffer;
            val += buffer;
        }
        val = -val;
        return val;